### DISCUSSION TITLE
Moving `Effect` to it's own Product or Package

### DISCUSSION THREAD

**Discussion by schrismartin** [pointfreeco/swift-composable-architecture#838] [Ideas]
My team _loves_ using TCA to build new features within our apps. 

Our project has a large set-up with a number of internally maintained repositories managed through SPM that we use to encapsulate individual features, components, network libraries, etc. 

For those micro-frameworks that don't warrant a complete import of TCA, we'd love to use the `Effect` type - because of its simplicity of use, ergonomics (like `.fireAndForget` and `future`), consistency across all of our frameworks that use Combine to represent side-effects, and the need to suffix all sorts of things with `eraseToEffect` within TCA. 

Ideally, we'd love to see this as another repo (with decoupled versioning from TCA) so that we can import it without the cruft of TCA, but it would still be useful if the `Effect` was in its own module and vended as a separate product of the TCA package.

Thanks for your consideration! Cheers! 🍻

**Comment by stephencelis** [comment]
@schrismartin Thanks for the discussion!

We definitely see the appeal of breaking this type out to its own library, though we have been hesitant for a few reasons. Let us know what you think:

First, we think you can get a lot of mileage out of vanilla Combine if you simply want to produce a library of effects and not depend on TCA. Your library can vend some `AnyPublisher`s and then in TCA you can `.eraseToEffect()` or `.catchToEffect()` in your reducer, which is usually necessary anyway to bundle up the output into a reducer action. There are a few useful helpers that TCA provides, as you mention, but many of them can be recreated in a line or two of vanilla Combine.

Second, with an eye to the future, Swift's async story is finally coming to light, and we'd like to be on a road to minimize `Effect`'s surface area and maybe even eliminate it entirely some day. Ideally a reducer could return a generic async sequence, though it'll probably still be some time before Swift and TCA can support this.

We're definitely open to discussion here, and further thought, but we also don't want to make a change here without a lot of thought 😄

**Reply by schrismartin** [reply]
I totally understand the hesitancy to go down that road when async/await is maturing, especially given `Effect`'s close ties with `Combine`.
